flux-jobspec command

SteVwonder opened the pull request:
As promised today over coffee, here is what I have so far.  It currently supports several flags that are common amongst `sbatch` and `srun`.  Including `-N` (nodes), `-c` (cores), `-n` (tasks), `--time`, and `--output`.  I don't think there is anything else we need for Corona other than `-g` (gpus) and `--error`.

Oh, it also needs to actually do something when someone supplies `--output` (whoops, forgot that one).

Related #1929 (still needs oar support before closing it)
Closes #2003

SteVwonder commented:
Currently when a user provides `--time`, it gets validated against the SLURM walltime format and then the string the user provided at the command line is added straight into the jobspec without any modification.  If we have a specific format for the `duration` in the jobspec that we want to use, it would be really easy to convert the SLURM format into something else.  We just need to make a decision on that.

garlick commented:
This is great - both instructive and useful for testing.

Maybe no big deal if it's already done, but
* Given this is _plumbing_, `jq` could perform any prettification on the pipeline, and the `--pretty` and `--compact` options could be dropped.
* I would vote for `srun` only, leave off sbatch since srun seems to be a superset?
* no need to go nuts with the distributions IMHO since we can't do much with them in the new system yet

I'd like to see this merged sooner rather than later.  Nice work :-)

grondo commented:
This doesn't have to be done before this PR is merged (the current version is quite useful as it is), however it would be nice if eventually the `flux-jobspec` command had an interface similar to other `flux` subcommands (as much as possible), e.g. treat `srun` or `sbatch` or `whatever` as sub-subcommands, not a "choice".

Also `flux jobspec help` should print the help for the `jobspec` command, including list of subcommands, and `flux jobspec srun --help` should print help for just the `flux jobspec srun` subcommand.

E.g. from `flux content`:

```
$ flux content help
flux content: Unknown subcommand: help
Usage: flux content dropcache [OPTIONS]
   or: flux content flush [OPTIONS]
   or: flux content load [OPTIONS] BLOBREF
   or: flux content spam N [M]
   or: flux content store [OPTIONS]
Access content store
  -h, --help             Display this message
$ flux content spam --help
Usage: flux content spam N [M]
Store N random entries, keeping M requests in flight (default 1)
  -h, --help             Display this message.
```

It probably seems minor, but having all the `flux blah` commands behave similarly will reduce user frustration in the long run I think.

I wonder if we could subclass the python `argparse` library to make this happen easily for this and future python commands?

On `--yaml` vs `--json`, would it be simpler and more extensible to just have a `--output-format`, `--of` or similar option? I'm actually not sure it matters either way, but it seems a more tractable way to add other output formats in the future (if that would even be needed)

dongahn commented:
>It probably seems minor, but having all the flux blah commands behave similarly will reduce user frustration in the long run I think.

> I wonder if we could subclass the python argparse library to make this happen easily for this and future python commands?

@grondo and @SteVwonder: `flux resource` would be an example which can benefit from this. 

https://github.com/flux-framework/flux-sched/blob/master/t/scripts/flux-resource

But given that Python did a fair good job of abstracting this out. Do we want to abstract this on our own, though.

grondo commented:
> But given that Python did a fair good job of abstracting this out. Do we want to abstract this on our own, though.

I don't think we should rewrite `argparse`, but I was hoping the module was flexible enough to at least customize the default help and other output to match existing flux commands.

dongahn commented:
> I don't think we should rewrite argparse, but I was hoping the module was flexible enough to at least customize the default help and other output to match existing flux commands.

I see. Maybe flux-resource can be another data point then. 

I'm not a python expert but I think it was flexible enough when I used it. I was at least trying to micmic flux commands/subcommands. Do you think it matches with what you had in mind or lacks some important features?

SteVwonder commented:
> Given this is plumbing, jq could perform any prettification on the pipeline, and the --pretty and --compact options could be dropped.

Done (push coming shortly)

> I would vote for srun only, leave off sbatch since srun seems to be a superset?

Done

> no need to go nuts with the distributions IMHO since we can't do much with them in the new system yet

Ok.  Should I pull that out as well, or just leave it in as-is?

> Also flux jobspec help should print the help for the jobspec command, including list of subcommands, and flux jobspec srun --help should print help for just the flux jobspec srun subcommand.
> I wonder if we could subclass the python argparse library to make this happen easily for this and future python commands?
> I don't think we should rewrite argparse, but I was hoping the module was flexible enough to at least customize the default help and other output to match existing flux commands.

I played around with this for a few hours last night.  Unfortunately, argparse does not make it easy to customize the output of the usage/help.  I opened a ticket to track that discussion/issue further: #1993

grondo commented:
> > I would vote for srun only, leave off sbatch since srun seems to be a superset?

> Done

I didn't have a chance to look at the initial implementation, but isn't the key difference between `flux jobspec srun` and `sbatch` that `sbatch` should insert a `flux start` or `flux broker` before the provided `COMMAND` (which is assumed to be a batch script), and ensure that the tasks run once per node? (or something similar

SteVwonder commented:
> isn't the key difference between flux jobspec srun and sbatch that sbatch should insert a flux start or flux broker before the provided COMMAND (which is assumed to be a batch script), and ensure that the tasks run once per node

My implementation is/was not that advanced :laughing: They were just aliases, basically.  That is a really good idea though.  To make that last piece about one per node work, I think `--nodes` would have to be a required argument.  If it is optional and the user only specifies `--cores`, then the tool will have to make a determination/assumption about the number of cores per node......actually, now that I think about this, is it even possible to specify that kind of request in the canonical jobspec? The request of thinking of is along the lines of: "I want N cores. I don't care about the number of nodes, but I want 1 process per node."

grondo commented:
> The request of thinking of is along the lines of: "I want N cores. I don't care about the number of nodes, but I want 1 process per node.

Good question! This was much easier to reason about when the `slot:` could appear along with the task intead of the resource section. :expressionless:

grondo commented:
It does feel like we got ourselves into a bind by requiring that the `slot` shape be defined as part of the resource request. There are certain things that are now not possible or very difficult to express (though it could just be my misunderstanding of current jobspec definition). In general, I don't see how you can define a slot at a parent of the resource type you want to request without affect the exact count (and exclusivity) of the parent resource type (if that makes any sense, specific example is what we're discussing here where you want N cores with a slot of a node, but don't care how many nodes. In fact, the exact slot shape can only be determined from a *concrete* resource set)

A couple ideas on how to handle this come to mind:

 * add a concept of predefined slots that match resource names, e.g. `node`, `socket`, `core` that allow the slot to be left out of the jobspec resources section, e.g. in the case of `flux jobspec sbatch` you could emit a resources section with no slot and a `task:` with `slot:` label `node:`

 * add a special value for `count:` like `any` or `0` that indicates no preference for the number of this type, and any slot that contains an `any` resource as a direct child would by definition apply to exactly 1 of that resource.

I'll be embarrassed if we've already discussed this somewhere and I missed it...

SteVwonder commented:
> I'll be embarrassed if we've already discussed this somewhere and I missed it...

I'm not seeing it anywhere, so I'll open a new issue and copy over the conversation so far.

SteVwonder commented:
This PR is ready for another review.  Thanks @grondo and @trws for whiteboarding the different situations with me on Friday.  That made writing tests super simple.

I opted to remove the `--distribution` flag.  It didn't seem necessary after our conversation on Friday.  I also switched from `--yaml` and `--json` to `--format yaml` and `--format json` as @grondo suggested above.

garlick commented:
Awesome!

One minor thing is that because this script lacks the `.py` suffix, it's not subject to the black format check (and it finds some small nits).  I'm not sure what we should do about that - have the `check-format` script also look for scripts containing a python shebang?   I'll open a separate issue on that but the formatting should probably be adjusted here.

SteVwonder commented:
Great catch @garlick!

Rebased, formatter applied to `src/cmd/flux-jobspec` at relevant commits, and added a fix to `scripts/{check-,}format` so that they run on `src/cmd/flux-jobspec` (see #2003).  I can pull that last addition out pretty easily if we want to go a different route.

grondo commented:
Really nice!

Just a couple very minor things while playing around with this really nice tool:
 
Errors should probably be prefixed with `flux-jobspec: ` so that in a pipeline or scripted invocation, the user knows from whence the error came. Actually I just noticed some errors do have `flux-jobspec` prefix, others do not:
 ```
$ flux jobspec srun -n0
usage: flux-jobspec [-h] [--format {json,yaml}] {srun} ...
flux-jobspec: error: command is required
$ flux jobspec srun -n0 hostname
Unknown error: resource count must be > 0
$ flux jobspec srun -n1 -N2 hostname
Input Error: Number of nodes greater than the number of tasks
 ```

I'm probably confused from our prior discussion, but this one doesn't seem right. I want to run 2 tasks with 1 core per task (implied) on a single node (is this just a need to special case nnodes == 1?)

```
$ flux jobspec srun -n2 -N1 hostname | jq -S .
Number of tasks is not an integer multiple of the number of nodes. More resources than required will be requested to satisfy your job
{
  "attributes": {
    "system": {}
  },
  "resources": [
    {
      "count": 1,
      "type": "node",
      "with": [
        {
          "count": 2,
          "label": "task",
          "type": "slot",
          "with": [
            {
              "count": 1,
              "type": "core"
            }
          ]
        }
      ]
    }
  ],
  "tasks": [
    {
      "attributes": {},
      "command": [
        "hostname"
      ],
      "count": {
        "total": 2
      },
      "slot": "task"
    }
  ],
  "version": 1
}
```

SteVwonder commented:
> Errors should probably be prefixed with flux-jobspec: so that in a pipeline or scripted invocation, the user knows from whence the error came.

Good catch!  I switched from using manual print statements to using the python `logging` module.  How does this look?
```
$ ./flux jobspec srun -n0
usage: flux-jobspec [-h] [--format {json,yaml}] {srun} ...
flux-jobspec: error: command is required
$ ./flux jobspec srun -n0 hostname
flux-jobspec: ERROR: resource count must be > 0
$ ./flux jobspec srun -n1 -N2 hostname
flux-jobspec: ERROR: Number of nodes greater than the number of tasks
$ ./flux jobspec srun -n14 -N2 hostname
flux-jobspec: WARNING: Number of tasks is not an integer multiple of the number of nodes. More resources than required will be requested to satisfy your job.
{"tasks": [{"slot": "task", "count": {"total": 14}, "command": ["hostname"], "attributes": {}}], "attributes": {"system": {}}, "version": 1, "resources": [{"count": 2, "with": [{"count": 7, "with": [{"count": 1, "type": "core"}], "type": "slot", "label": "task"}], "type": "node"}]}
```

Any errors from the `argparse` parser will have a lowercase `error` and print out the usage, while exceptions thrown during validation or program execution will have the all caps `ERROR`.  I vote to mess around with the exact formatting of the `argparse` errors in #1993.

> I'm probably confused from our prior discussion, but this one doesn't seem right.

Great catch!  I had my modulo test backwards.  I was testing `num_nodes % num_tasks == 0` when I should have been testing `num_tasks % num_nodes == 0`.  I flipped it and added a test to check that:
- no warning is printed on a correct invocation (i.e., -N8 -n16)
- num_tasks that are integer multiples of num_nodes produce a jobspec that uses `.tasks[0].count.per_slot` rather than `.tasks[0].count.total`.

End result:
```yaml
$ ./flux jobspec srun -n2 -N1 hostname | jq -S .    
{
  "attributes": {
    "system": {}
  },
  "resources": [
    {
      "count": 1,
      "type": "node",
      "with": [
        {
          "count": 2,
          "label": "task",
          "type": "slot",
          "with": [
            {
              "count": 1,
              "type": "core"
            }
          ]
        }
      ]
    }
  ],
  "tasks": [
    {
      "attributes": {},
      "command": [
        "hostname"
      ],
      "count": {
        "per_slot": 1
      },
      "slot": "task"
    }
  ],
  "version": 1
}
```

garlick commented:
Sooo...is this one good to go after a rebase?

SteVwonder commented:
Rebased and did a little squashing.  I think this is good to go once Travis is green.

garlick commented:
Thanks - this is a high value one!
